java-analyzer-bundle.core/src/main/java/io/konveyor/tackle/core/internal/symbol/TypeSymbolProvider.java

```diff
@@ -5,27 +5,95 @@
 import java.util.ArrayList;
 import java.util.List;
 
+import org.eclipse.core.runtime.URIUtil;
+import org.eclipse.jdt.core.IClassFile;
+import org.eclipse.jdt.core.ICompilationUnit;
+import org.eclipse.jdt.core.IImportDeclaration;
 import org.eclipse.jdt.core.IJavaElement;
+import org.eclipse.jdt.core.IPackageDeclaration;
 import org.eclipse.jdt.core.search.SearchMatch;
 import org.eclipse.jdt.core.search.TypeDeclarationMatch;
 import org.eclipse.jdt.core.search.TypeParameterDeclarationMatch;
 import org.eclipse.jdt.core.search.TypeParameterReferenceMatch;
 import org.eclipse.jdt.core.search.TypeReferenceMatch;
+import org.eclipse.lsp4j.Location;
 import org.eclipse.lsp4j.SymbolInformation;
 import org.eclipse.lsp4j.SymbolKind;
 
-public class TypeSymbolProvider implements SymbolProvider {
+public class TypeSymbolProvider implements SymbolProvider, WithQuery {
+    private String query;
+
     @Override
     public List<SymbolInformation> get(SearchMatch match) {
         SymbolKind k = convertSymbolKind((IJavaElement) match.getElement());
         List<SymbolInformation> symbols = new ArrayList<>();
         // For Method Calls we will need to do the local variable trick
-        if (!(match instanceof TypeReferenceMatch || 
-            match instanceof TypeDeclarationMatch || 
-            match instanceof TypeParameterDeclarationMatch || 
-            match instanceof TypeParameterReferenceMatch)) {
-                return null;
+        if (!(match instanceof TypeReferenceMatch ||
+                match instanceof TypeDeclarationMatch ||
+                match instanceof TypeParameterDeclarationMatch ||
+                match instanceof TypeParameterReferenceMatch)) {
+            return null;
+        }
+        // TypeReferenceMatch often are inaccurate in that if a pattern is a.b.C we get
+        // matches for references of C no matter whether they are in package a.b or not
+        // we try to confirm whether C we are getting belongs to package a.b with checks:
+        // first, we check if the file belongs to package a.b
+        // second, we check if CompilationUnit has package declaration of a.b
+        // third, we check if CompilationUnit has explicit import of a.b.C or a.b.*
+        if (match instanceof TypeReferenceMatch && this.query.contains(".")) {
+            try {
+                String qualification = "";
+                int dotIndex = this.query.lastIndexOf('.');
+                if (dotIndex > 0) {
+                    qualification = this.query.substring(0, dotIndex);
+                }
+                var element = (IJavaElement) match.getElement();
+                ICompilationUnit compilationUnit = (ICompilationUnit) element
+                        .getAncestor(IJavaElement.COMPILATION_UNIT);
+                if (compilationUnit == null) {
```

Comment:
Just for my own understanding when does this happen?

Reply:
If I understood it correctly, this happens when match is in  a .class file. In that case, the compilation unit needs to be loaded into memory. Fortunately, the class file doesn't need to be decompiled....we can somehow get imports, package declaration, methods etc just from the class file handle